cirq/circuits/quil_output.py

```diff
@@ -0,0 +1,224 @@
+
+    def __init__(self, operations: 'cirq.OP_TREE',
+                 qubits: Tuple['cirq.Qid', ...]) -> None:
+        """
+        Args:
+            operations: A list or tuple of `cirq.OP_TREE` arguments.
+            qubits: The qubits used in the operations.
+        """
+        self.qubits = qubits
+        self.operations = tuple(cirq.ops.flatten_to_ops(operations))
+        self.measurements = tuple(op for op in self.operations
+                                  if isinstance(op.gate, ops.MeasurementGate))
+        self.qubit_id_map = self._generate_qubit_ids()
+        self.measurement_id_map = self._generate_measurement_ids()
+        self.formatter = protocols.QuilFormatter(
+            qubit_id_map=self.qubit_id_map,
+            measurement_id_map=self.measurement_id_map)
+
+    def _generate_qubit_ids(self) -> Dict['cirq.Qid', str]:
+        return {qubit: str(i) for i, qubit in enumerate(self.qubits)}
+
+    def _generate_measurement_ids(self) -> Dict[str, str]:
+        index = 0
+        measurement_id_map: Dict[str, str] = {}
+        for op in self.operations:
+            if isinstance(op.gate, ops.MeasurementGate):
+                key = protocols.measurement_key(op)
+                if key in measurement_id_map:
+                    continue
+                measurement_id_map[key] = 'm{}'.format(index)
+                index += 1
+        return measurement_id_map
+
+    def save_to_file(self, path: Union[str, bytes, int]) -> None:
+        """Write QUIL output to a file specified by path."""
+        with open(path, 'w') as f:
+
+            def write(s: str) -> None:
+                f.write(s)
+
+            self._write_quil(write)
+
+    def __str__(self) -> str:
+        output = []
+        self._write_quil(lambda s: output.append(s))
+        return self.rename_defgates(''.join(output))
+
+    def _write_quil(self, output_func: Callable[[str], None]) -> None:
+        output_func('# Created using Cirq.\n\n')
+        if len(self.measurements) > 0:
+            measurements_declared: Set[str] = set()
+            for m in self.measurements:
+                key = protocols.measurement_key(m)
+                if key in measurements_declared:
+                    continue
+                measurements_declared.add(key)
+                output_func('DECLARE {} BIT[{}]\n'.format(
+                    self.measurement_id_map[key], len(m.qubits)))
+            output_func('\n')
+
+        def keep(op: 'cirq.Operation') -> bool:
+            print("I'm keeping")
+            return protocols.quil(op, formatter=self.formatter) is not None
+
+        def fallback(op):
+            if len(op.qubits) not in [1, 2]:
+                return NotImplemented
+
+            mat = protocols.unitary(op, None)
+            if mat is None:
+                return NotImplemented
+
+            if len(op.qubits) == 1:
+                return QuilOneQubitGate(mat).on(*op.qubits)
+            return QuilTwoQubitGate(mat).on(*op.qubits)
```

Comment:
I've been trying to find a way to cover the top three lines. It seems like all of the gates decompose into gates that we have defined `_quil_` functions for. My thoughts are to make a basic custom one and two qubit gate that wouldn't decompose but I haven't been able to figure out a good way to do that. @vtomole Do you know of anywhere in the code I could find an example that may have done something similar?

Reply:
`qasm_output_test.py` doesn't help?

Reply:
Correct. `qasm_output_test.py` doesn't help because QASM had basic gates that were not implemented. Whereas with QUIL, all gates I have found decompose into QUIL standard gates which do not require the `fallback` function from `protocols.decompose`

Reply:
I'm going to merge and file an issue for this.